ugs-platform/ugs-platform-visualizer/src/main/java/com/willwinder/ugs/nbm/visualizer/renderables/MachineBoundry.java

```diff
@@ -0,0 +1,180 @@
+package com.willwinder.ugs.nbm.visualizer.renderables;
+
+import com.jogamp.opengl.GL2;
+import com.jogamp.opengl.GLAutoDrawable;
+import com.willwinder.ugs.nbm.visualizer.options.VisualizerOptions;
+import com.willwinder.ugs.nbm.visualizer.shared.Renderable;
+import com.willwinder.ugs.nbp.lib.lookup.CentralLookup;
+import com.willwinder.universalgcodesender.firmware.FirmwareSettingsException;
+import com.willwinder.universalgcodesender.model.BackendAPI;
+import com.willwinder.universalgcodesender.model.UGSEvent;
+
+import javax.vecmath.Point3d;
+
+import static com.jogamp.opengl.GL.GL_LINES;
+import static com.willwinder.ugs.nbm.visualizer.options.VisualizerOptions.VISUALIZER_OPTION_X;
+import static com.willwinder.ugs.nbm.visualizer.options.VisualizerOptions.VISUALIZER_OPTION_Y;
+import static com.willwinder.ugs.nbm.visualizer.options.VisualizerOptions.VISUALIZER_OPTION_Z;
+
+public class MachineBoundry extends Renderable {
+    private final BackendAPI backendAPI;
+    private double softLimitX = 0;
+    private double softLimitY = 0;
+    private double softLimitZ = 0;
+    private float[] machineBoundryBottomColor;
+    private float[] machineBoundryLineColor;
+    private float[] yAxisColor;
+    private float[] xAxisColor;
+    private float[] zAxisColor;
+    private boolean softLimitsEnabled;
+
+    /**
+     * Construct with a priority number. Objects should be rendered from highest
+     * to lowest priority.
+     */
+    public MachineBoundry() {
+        super(Integer.MIN_VALUE, "Machine boundries");
+        reloadPreferences(new VisualizerOptions());
+        backendAPI = CentralLookup.getDefault().lookup(BackendAPI.class);
+        backendAPI.addUGSEventListener(this::onUGSEvent);
+    }
+
+    private void onUGSEvent(UGSEvent event) {
+        try {
+            softLimitsEnabled = backendAPI.isConnected() && backendAPI.getController().getFirmwareSettings().isSoftLimitsEnabled();
+            if (backendAPI.isConnected() && backendAPI.isIdle() && backendAPI.getController().getCapabilities().hasSoftLimits() && softLimitsEnabled) {
```

Comment:
Is the idle check necessary? I'm also wondering about the `hasSoftLimits` `isSoftLimitsEnabled` combination. Maybe `GrblFirmwareSettings` could be initialized with the capabilities object so that it can do the check implicitly?

Reply:
I'd also like to favor the `getSoftLimit(X/Y/Z)` methods, I think that would be better in case additional axes are added.